### ISSUE TITLE
Authentication failed - dt prace account

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#542]
In GitLab by @ mariag on Jun 9, 2020, 11:14

Hello all, 

I have a similar issue to #536 and I cannot find a solution.

I've created an auto-ecearth experiment as copy of a bsc32-mn4 one. Then I've changed the settings to use the PRACE account. The expid is **a2wq**. 

The information I've got in the .ssh/config is: 

```
 # prace 
 Host mn4_prace 
 Hostname mn3.bsc.es
 User pr1eld06
 ForwardAgent yes
 IdentityFile ~/.ssh/id_rsa_prace
 ForwardX11 yes
 # prace_dt 
 Host dt_prace 
 Hostname dt02.bsc.es
 User pr1eld06
 IdentityFile ~/.ssh/id_rsa_prace
 ForwardX11 yes
```

The platforms_a2wq.conf has:

```
[transfer_node]
TYPE = ps
HOST = dt_prace
PROJECT = pr1eld00
USER = pr1eld06
ADD_PROJECT_TO_HOST = false
SCRATCH_DIR = /gpfs/scratch
```

Then I have as jobs: 
```
[LOCAL_SETUP]
FILE = templates/common/localsetup.tmpl.sh
PLATFORM = LOCAL
CHECK = true
NOTIFY_ON = FAILED

[SYNCHRONIZE]
FILE = templates/common/synchronize.tmpl.sh
PLATFORM = transfer_node
DEPENDENCIES = LOCAL_SETUP
CHECK = true
NOTIFY_ON = FAILED
```

When I try to run the experiment, it crashes due to an authentication error (log in: /esarchive/autosubmit/a2wq/runs/nohup.out )
```
[ERROR] transfer_node submission failed
[CRITICAL] Unhandled exception on Autosubmit: Authentication failed.
```

I've created the platforms_a2wq.conf file from scratch again to test if there were some hidden characters there causing the issue. Also, I've used another experiment to check and the authentication failure still happens. 

If I do ssh dt_prace from my WS and the AS machines it connects correctly to the prace account. 

Any suggestion?

Thanks!!

FYI: @dbeltrankyl @gmontanepinto

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:11

Does `\~/.ssh/id_rsa_prace` still exist?

If that's the case it seems Autosubmit is using the default, because you didn't change id_rsa_prace

**Comment by kinow** [comment]
In GitLab by @ mariag on Jun 9, 2020, 15:12

Nice, @mcastril ! It exists! :) 

So, how do I tell autosubmit to use id_rsa_prace?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:14

We think that it can be a bug and that maybe Autosubmit is not able to use it. But it is strange because in the case of #536 it seems it was working.

Marc had a '/' at the end... This would be very weird but you maybe can try it.

```
//  IdentityFile ~/.ssh/id_rsa_prace_kim//
//  ForwardX11 yes//
```

On the other hand, @dbeltrankyl , you maybe can do a basic test, of renaming your own id_rsa file and using it in your `.ssh/config` . And then see what Autosubmit is doing.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:17

@dbeltrankyl , for Paramiko it seems small letters is the way to do it:

https://gist.github.com/acdha/6064215#gistcomment-1779301

**Comment by kinow** [comment]
In GitLab by @ mariag on Jun 9, 2020, 15:20

This gets weirder and weirder, now it understands that the IdentityFile is a directory? 

```
Jobs ready for transfer_node: 1
[ERROR] Can not create ssh connection to dtprace: Not a directory
WARNING:paramiko.transport:Oops, unhandled type 3
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:36

Ok so it seems that in the case of @ mguevara it was indeed a directory, but of course it is not needed.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 9, 2020, 15:39

hi, I did some tests:

The variable is well written, once in the code it automatically is set in minus independently if it had caps on the .ssh/config file


The variable is getting correctly, however, it for some reason it is getting an incorrect variable type. Concretely is getting a list instead of the expected string (at least as paramiko docs specifies).

However, once in the paramiko client code, it indeed is using the correct id_rsa for my case.

`for key_filename in key_filenames` so the type seems the indicated although the doc say a different thing.

Also the order to find the keys is:
```
         Try, in order:
             - The key passed in, if one was passed in.>             
             - Any key we can find through an SSH agent (if allowed).
             - Any "id_rsa", "id_dsa" or "id_ecdsa" key discoverable in ~/.ssh/
               (if allowed).
             - Plain username/password auth, if a password was given.
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:52

Here it could also happen that you only copied your public key and not the private, and there is a mismatch.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 9, 2020, 15:55

Now I created a new key, having two and tested the correct one and then the wrong one. 

Both works. But there is a why.

If the specified one auth doesn't work, it will fallback into the .ssh/id_rsa one and beginthe run with that one instead of the specified

So with that in mind, I think the error is the key itself, I recommend to you to regenerate the prace key and upload the key again to the hpc portal. @ mariag.

Basically, it follows the order specified in the previous commentary, and when one works autosubmit(paramiko) will run with that one.

In my case, it worked for
* 1) option for the correct one
* 3) option for the incorrect one.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:58

> Now I created a new key, having two and tested the correct one and then the wrong one. 

This was an Autosubmit test?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 9, 2020, 15:58

yes

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:58

But she said she can login with that key in PRACE, so the key should be fine no?

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 15:59

So you suggest that maybe the key is fine for the manual ssh connection but Paramiko is finding a problem and trying to use the default one.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jun 9, 2020, 15:59

Yes, but maybe ssh remember the password in a cache.. or it is using bsc32 key? I don't really know but otherwise, I'm totally lost

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 16:00

This would explain this error:

```
[ERROR] transfer_node submission failed
[CRITICAL] Unhandled exception on Autosubmit: not a valid RSA private key file
```

But there is a possibility that this is because she renamed the public but not the private key to do the test.

**Comment by kinow** [comment]
In GitLab by @ mariag on Jun 9, 2020, 16:06

Eh, I renamed both :) 

To sum up:

I'll regenerate the keys to PRACE, upload them to the portal and remove the last / I added to test. 

I'll keep you updated.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 16:11

Thanks @ mariag

**Comment by kinow** [comment]
In GitLab by @ mariag on Jun 9, 2020, 16:52

Don't ask me why, but the new keys work. 

```
82 of 82 jobs remaining (16:50)
Job a2wq_LOCAL_SETUP is COMPLETED
Jobs ready for transfer_node: 1
a2wq_SYNCHRONIZE submitted


81 of 82 jobs remaining (16:50)
Job a2wq_SYNCHRONIZE is RUNNING
```

Could it be because I generated the last ones from the WS and these from the AS machine?? 

In any case, many thanks to both of you @dbeltrankyl and @mcastril

**Comment by kinow** [comment]
In GitLab by @mcastril on Jun 9, 2020, 16:58

Good that it worked!

My guess is that something was wrong with the previous keys. Are you still using `IdentityFile ~/.ssh/id_rsa_prace` ?

**Comment by kinow** [comment]
In GitLab by @ mariag on Jun 9, 2020, 17:01

Yes, same as before. Just changed the keys to the new ones and upload them again to the HPC portal... 

I'll close the issue. 

Thanks a lot for your time, guys!

**Comment by kinow** [comment]
In GitLab by @ mariag on Jun 9, 2020, 17:01

closed
